src/merkle_tree.rs

```diff
@@ -72,9 +72,24 @@ where
         }
     }
 
-    pub fn nodes(&self) -> &Vec<T> {
+    pub fn nodes(&self) -> &[T] {
         &self.nodes
     }
+
+    /// Verifies that a proof points to leaves in a tree, returning an empty vector if the proof can not be verified
+    pub fn verify(&self, proof: MerkleProof<T, M>) -> Vec<T> {
```

Comment:
This interface is very strange and misleading.

Typically, a verifier doesn't have enough data to construct a Merkle tree. Instead of holding a whole tree, a verifier only has a part of the data: `leaves` list and `MerkleProof`.

This interface adds a method on MerkleTree, which implies that the verifier should build the MerkleTree first, then to verifies the merkle proof, it is not appropriate for the real-world scenario.

Reply:
it's from the requirement of this real case:
https://bitcoincore.org/en/doc/0.18.0/rpc/blockchain/verifytxoutproof/

Reply:
Can also see as:
- check if the proof is belong to the tree
- if it belongs to the tree return the leaves which build the proof

Reply:
This RPC seems can be built outside the `merkle-tree` structure: to collect the txs from the block then verifies the merkle_root.

I don't see any reason to put this function into a `merkle-tree` crate, it is a very specific requirement of Bitcoin and CKB; and it can be implemented without adding a new function.

Reply:
By using this function to implement the RPC, we need to load all txs in the block to build the merkle tree. But actually we only need to load a part of txs and do the merkle verification, this interface is inefficient for the purpose. I suggest we could implement this RPC without touch the `merkle-tree` crate.

Reply:
I think you misunderstand the purpose of this function, it is used when the person holding the proof does not have the corresponding elements for indices, he has no way to verify the proof,
he can delegate the verification to a third party who holding the full data (tree) to help him to retrieve corresponding elements.

In your case, `MerkleProof#verify(root: &T, leaves: &[T])` should be used.

Reply:
I don't think I misunderstand the purpose.

1. The person(verifier) requests a full node(prover) to retrieve the elements.
2. The prover compute the indexes of txs by reverse the calculation https://github.com/nervosnetwork/merkle-tree/blob/a4ee84f63dd2b5158228724e40aebc8786992baf/src/merkle_tree.rs#L32
3. The prover retrieves a part of txs which match the proof and send back to the verifier.
4. The verifier does the merkle verification.

This process does not involve this new `verify` function. If you must use the `verify` function to implement the feature, in step 2 the prover must load all txs of a block which is inefficient than the description above.

Reply:
I agree that what we need is just the way to get leaf hashes by index. From the proof, it is easy to convert the node indices to leaf indices.

Reply:
@jjyr @doitian a fn which converts the node indices to leaf indices isn't enough, the prover still need to re-construct a tree and calculate the tree root to verify it matches proof's root, that's the reason I add this fn to tree struct, please review it again, thanks.

Reply:
In application, the root usually is already known, or has a fast way to check it instead of reconstructing the whole tree.

Reply:
BTW, I also suggest **even** we decide to implement the RPC in this way, it is better to put the code in the CKB or light node side rather than put it in the merkle-tree crate since it is a specific function for implement light protocol(merkle tree crate already exports all necessary functions to implement this).

Reply:
> In application, the root usually is already known, or has a fast way to check it instead of reconstructing the whole tree.

As you said this fn is added for cases where no root is saved and the prover needs to rebuild the entire tree.

Reply:
Any provider can verify the proof if

* It knows count of leaves and the value of each leaf.
* It knows how to check the root in proof matches.

Say the tree has 1024 leaves and the proof has one leaf, and the root is already known by the proof.

* A simple method can verify the proof with only reading one leaf and merging 10 times.
* While rebuilding the tree requires extra 1024 reading and 1014 merges.

My concern of adding such method is that it is misleading that the only way to verify the proof is rebuilding the tree. My suggestion is adding a trait such as

```
trait ProofVerifier
    get_leaf_by_merkle_tree_node_index(i: usize) -> Option<T>;
    verify_root(root: T) -> bool;
```

Reply:
> BTW, I also suggest **even** we decide to implement the RPC in this way, it is better to put the code in the CKB or light node side rather than put it in the merkle-tree crate since it is a specific function for implement light protocol(merkle tree crate already exports all necessary functions to implement this).

it's just a helper fn for DRY principle,  if your point is valid, why we still provide a helper fn `MerkleTree#verify` since this crate already exports the `MerkleTree#root`?

```
    pub fn verify(&self, root: &T, leaves: &[T]) -> bool {
        match self.root(leaves) {
            Some(r) => &r == root,
            _ => false,
        }
    }
```

Reply:
> My concern of adding such method is that it is misleading that the only way to verify the proof is rebuilding the tree. My suggestion is adding a trait such as

I got your point, this fn should be moved to `CBMT::retrieve_leaves` and use an optimized algorithm, if you are OK with this input and output, I will implement it like this:
```
/// Verifies that a proof points to leaves of a tree, returning `None` if the proof can not be verified
pub fn retrieve_leaves(leaves: &[T], proof: &MerkleProof) -> Option<Vec<T>> {
}
```

Reply:
Indeed the `retrieve_leaves` can be a method in the trait. The MerkleTree can implement the trait.

Reply:
CKB can implement the trait as:

* Get txids from the database or cache
* Verify that `merge(root, wtxids root)` equal to the root in the block header, where `wtxids root` must be an extra lemma in the proof for CKB